### PR TITLE
Add memkind shim library

### PR DESCRIPTION
**Pull request by michalbiesek**
- could be used with LD_PRELOAD mechanism for benchmarks created for standard malloc

### Types of changes

- [ ] Bugfix (non-breaking change which fixes issue linked in Description above)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)
- [ ] Other

### Checklist

- [ ] Code compiles without errors
- [ ] All tests pass locally with my changes (see TESTING section in CONTRIBUTING file)
- [ ] Created tests which will fail without the change (if possible)
- [ ] Extended the README/documentation (if necessary)
- [ ] All newly added files have proprietary license (if necessary)
- [ ] All newly added files are referenced in MANIFEST files (if necessary)

## Further comments

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/381)

### CODE REVIEW

**Review by pbalcer — COMMENTED** [review]
You should use malloc hooks, see:
https://github.com/pmem/vmem/blob/master/src/libvmmalloc/libvmmalloc.c

File: utils/memkind_malloc_wrapper/memkind_malloc_wrapper.c [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+static int is_dax_kmem_enabled(const char *env)
+{
+    if (env && strcmp(env, "1") == 0 &&
+        !memkind_check_available(MEMKIND_DAX_KMEM)) {
+        return 0;
+    }
+    return 1;
+}
+
+static void MEMKIND_WRAPPER_INIT init_memkind_function(void)
```
Comment (pbalcer) [inline utils/memkind_malloc_wrapper/memkind_malloc_wrapper.c:26]: you need a destructor.

**Review by kasiawasiuta — APPROVED** [review]
> Reviewed 4 of 7 files at r1, 5 of 5 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/381)* status: all files reviewed, 1 unresolved discussion (waiting on @jschmieg, @michalbiesek, and @PatKamin)

File: utils/memkind_malloc_wrapper/memkind_malloc_wrapper.c (hunk 1 above)
Comment (michalbiesek) [reply utils/memkind_malloc_wrapper/memkind_malloc_wrapper.c:44]: Done - destructor added

**Comment by michalbiesek** [comment]
> You should use malloc hooks, see:
> https://github.com/pmem/vmem/blob/master/src/libvmmalloc/libvmmalloc.c

Done added.

**Review by pbalcer — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/381)* status: 1 of 8 files reviewed, 3 unresolved discussions (waiting on @jschmieg, @kasiawasiuta, @michalbiesek, and @PatKamin)

___
*[utils/memkind_malloc_wrapper/memkind_malloc_wrapper.c, line 82 at r4](https://reviewable.io/reviews/memkind/memkind/381#-MJWbUFS50jNzxSoxxSM:-MJWbUFT6HT_SYtWrCRI:bqgsa79) ([raw file](https://github.com/memkind/memkind/blob/2ab1dfb1a04932f892e6086b4171db088dbc2251/utils/memkind_malloc_wrapper/memkind_malloc_wrapper.c#L82)):*
> ```C
> {
>     LOG("Finalize called\n");
>     MemkindInitDone = 0;
> ```

don't you need to destroy the kind?
___
*[utils/memkind_malloc_wrapper/memkind_malloc_wrapper.c, line 119 at r4](https://reviewable.io/reviews/memkind/memkind/381#-MJWbYYgAorOOmAFi6xw:-MJWbYYgAorOOmAFi6xx:b1fp42v) ([raw file](https://github.com/memkind/memkind/blob/2ab1dfb1a04932f892e6086b4171db088dbc2251/utils/memkind_malloc_wrapper/memkind_malloc_wrapper.c#L119)):*
> ```C
>         return NULL;
>     memkind_t kind = get_current_kind();
>     LOG("\t calloc: size %zu\n", size);
> ```

debug prints?
___
*[utils/memkind_malloc_wrapper/README.md, line 5 at r4](https://reviewable.io/reviews/memkind/memkind/381#-MJWbaBnC-CNk_0yGsvM:-MJWbaBnC-CNk_0yGsvN:bvptejg) ([raw file](https://github.com/memkind/memkind/blob/2ab1dfb1a04932f892e6086b4171db088dbc2251/utils/memkind_malloc_wrapper/README.md#L5)):*
> ```Markdown
> This is utils/memkind_malloc_wrapper/README.
> 
> libmemkind_malloc_wrapper.so is an interposer library for redirecting heap allocations to memkind.
> ```

Say here that this is for benchmarking/testing purposes only at the moment. And that fork is not supported.

**Review by pbalcer — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/381#-:-MJaN3_W8sG8zKEgygmh:bnfp4nl)* status: 0 of 9 files reviewed, all discussions resolved (waiting on @jschmieg, @kasiawasiuta, and @PatKamin)

**Pull request merged**